R/particle_filter.R

```diff
@@ -394,6 +330,156 @@ particle_filter <- R6::R6Class(
+    log_likelihood = NULL,
+    index_state = NULL,
+
+    initialize = function(pars, generator, data, steps, n_particles, n_threads,
+                          initial, index, compare, seed, save_history) {
+      model <- generator$new(data = pars, step = steps[[1L]],
+                             n_particles = n_particles, n_threads = n_threads,
+                             seed = seed)
+
+      if (!is.null(initial)) {
+        initial_data <- initial(model$info(), n_particles, pars)
+        if (is.list(initial_data)) {
+          steps <- particle_steps(steps, initial_data$step)
+          model$set_state(initial_data$state, initial_data$step)
+        } else {
+          model$set_state(initial_data)
+        }
+      }
+
+      index_data <- if (is.null(index)) NULL else index(model$info())
+      if (!is.null(index_data$run)) {
+        model$set_index(index_data$run)
+      }
+
+      if (save_history) {
+        len <- nrow(steps) + 1L
+        state <- model$state(index_data$state)
+        history_value <- array(NA_real_, c(dim(state), len))
+        history_value[, , 1] <- state
+        history_order <- matrix(seq_len(n_particles), n_particles, len)
+        self$history <- list(
+          value = history_value,
+          order = history_order,
+          index = index_data$state)
+        ## We also need to export this out because the *final state*
+        ## needs to know this.
+        self$index_state <- index_data$state
+      }
+
+      ## Constants
+      private$generator <- generator
+      private$pars <- pars
+      private$data <- data
+      private$steps <- steps
+      private$n_particles <- n_particles
+      private$n_threads <- n_threads
+      private$initial <- initial
+      private$index <- index
+      private$compare <- compare
+
+      ## Variable (see also history)
+      self$model <- model
+      self$state <- model$run(steps[[1]])
+      self$log_likelihood <- 0.0
+    },
+
+    step = function() {
+      if (self$complete) {
+        stop("The particle filter has reached the end of the data")
+      }
+      self$step_current <- self$step_current + 1L
+      self$complete <- self$step_current >= nrow(private$steps)
+      step <- self$step_current
+      save_history <- !is.null(self$history)
+
+      state <- self$model$run(private$steps[step, 2])
+      if (save_history) {
+        self$history$value[, , step + 1L] <-
+          self$model$state(self$history$index)
+      }
+
+      data <- private$data[[step]]
+      log_weights <- private$compare(state, self$state, data, private$pars)
+
+      if (is.null(log_weights)) {
+        self$state <- state
+        if (save_history) {
+          self$history$order[, step + 1L] <- seq_len(ncol(state))
+        }
+      } else {
+        weights <- scale_log_weights(log_weights)
+        self$log_likelihood <- self$log_likelihood + weights$average
+        if (self$log_likelihood == -Inf) {
+          self$complete <- TRUE
+          return(-Inf)
```

Comment:
This seems reasonable. Does this work with `sample` ok? In which case I think this probably also closes #75 (could you add a test for that and it can be closed here - one possibility is zero lambda for Poisson and observed > 0)

Reply:
This is just the same behaviour as before, so #75 still needs fixing. I think I discussed this with @edknock and sketched a partial solution